README.md
# Table of Contents

1. [Overview of AWS EC2 FPGA Development Kit](#overview-of-aws-ec2-fpga-development-kit)
    - [Developer Support](#developer-support)
    - [Development Flow](#development-flow)
    - [Development environments](#development-environments)
    - [FPGA Developer AMI](#fpga-developer-ami)
    - [FPGA Hardware Development Kit (HDK)](#hardware-development-kit-hdk)
    - [FPGA Software Development Kit (SDK)](#runtime-tools-sdk)
    - [Software Defined Development Environment](#software-defined-development-environment)
1. [Amazon EC2 F1 platform features](#amazon-ec2-f1-platform-features)
1. [Getting Started](#getting-started)
    - [Getting Familiar with AWS](#getting-familiar-with-aws)
    - [First time setup](#setting-up-development-environment-for-the-first-time)
    - [Quickstarts](#quickstarts)
    - [How To's](#how-tos)
1. [Documentation Overview](#documentation-overview)

# Overview of AWS EC2 FPGA Development Kit

AWS EC2 FPGA Development Kit is a set of development and runtime tools to develop, simulate, debug, compile and run hardware accelerated applications on [Amazon EC2 F1 instances](https://aws.amazon.com/ec2/instance-types/f1/).
It is distributed between this github repository and FPGA Developer AMI - [Centos](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ)/[AL2](https://aws.amazon.com/marketplace/pp/B08NTMMZ7X) provided by AWS with no cost of development tools.

⚠️ <b>NOTE:</b> The developer kit is supported for Linux operating systems only.

## Developer Support

Opening a [GitHub Issue](https://github.com/aws/aws-fpga/issues) is the preferred method to get support with the AWS FPGA Development Kit. In addition, the [FPGA Development re:Post Tag](https://repost.aws/tags/TAc7ofO5tbQRO57aX1lBYbjA/fpga-development) is available to find FPGA-related discussion topics from the AWS community of customers, AWS customer support, and the AWS FPGA development team.

## Development Flow
After creating an FPGA design (also called CL - Custom logic), developers can create an Amazon FPGA Image (AFI) and easily deploy it to an F1 instance. AFIs are reusable, shareable and can be deployed in a scalable and secure way.

![Alt text](hdk/docs/images/f1-Instance-How-it-Works-flowchart.jpg)

## Development Environments

| Development Environment | Description | Accelerator Language | Hardware Interface | Debug Options| Typical Developer                                                     |
| --------|---------|-------|---------|-------|-----------------------------------------------------------------------|
| Software Defined Accelerator Development using [Vitis](Vitis/README.md)/[SDAccel](SDAccel/README.md)| Development experience leverages an optimized compiler to allow easy new accelerator development or migration of existing C/C++/openCL, Verilog/VHDL to AWS FPGA instances | C/C++/OpenCL, Verilog/VHDL (RTL) | OpenCL APIs and XRT | SW/HW Emulation, Simulation, GDB, Virtual JTAG (Chipscope) | SW or HW Developer with zero FPGA experience                          |
| [Hardware Accelerator Development using Vivado](hdk/README.md) | Fully custom hardware development experience provides hardware developers with the tools required for developing AFIs for AWS FPGA instances  | Verilog/VHDL | [XDMA Driver](sdk/linux_kernel_drivers/xdma/README.md), [peek/poke](sdk/userspace/README.md) | Simulation, Virtual JTAG | HW Developer with advanced FPGA experience                            |
| [IP Integrator/High Level Design(HLx) using Vivado](hdk/docs/IPI_GUI_Vivado_Setup.md) | Graphical interface development experience for integrating IP and high level synthesis development | Verilog/VHDL/C | [XDMA Driver](sdk/linux_kernel_drivers/xdma/README.md), [peek/poke](sdk/userspace/README.md) | Simulation, Virtual JTAG | HW Developer with intermediate FPGA experience                        |
 | [On-premise development for Alveo U200 using Vitis targetted for migration to F1](Vitis/docs/Alveo_to_AWS_F1_Migration.md) | Vitis flow development using on-premise U200 platform targeted for migration to F1 |  C/C++/OpenCL, Verilog/VHDL (RTL) | OpenCL APIs and XRT | SW/HW Emulation, Simulation, GDB, JTAG (Chipscope) | SW or HW Developer with zero FPGA experience and on-premise U200 card |
 | [On-premise development for Alveo U200 using F1.A.1.4 shell](hdk/docs/U200_to_F1_migration_HDK.md) | HDK flow for on-premise U200 card using F1.A.1.4 shell targetted for migration to F1 | Verilog/VHDL | XDMA driver, peek/poke | Simulation, JTAG | HW Developer with advanced FPGA experience and on-premise U200 card   |
> For on-premise development, SDAccel/Vitis/Vivado must have the [correct license and use one of the supported tool versions](./docs/on_premise_licensing_help.md). 

## FPGA Developer AMI

The [FPGA Developer AMI](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ) is available on the AWS marketplace without a software charge and includes tools needed for developing FPGA Designs to run on AWS F1. 

Given the large size of the FPGA used inside AWS F1 Instances, Xilinx tools work best with 32GiB Memory. 
z1d.xlarge/c5.4xlarge and z1d.2xlarge/c5.8xlarge instance types would provide the fastest execution time with 30GiB+ and 60GiB+ of memory respectively. 
Developers who want to save on cost, could start coding and run simulations on low-cost instances, like t2.2xlarge, and move to the aforementioned larger instances to run the synthesis of their acceleration code.

AWS marketplace offers multiple versions of the FPGA Developer AMI. The following section table describes the mapping of currently supported developer kit versions to AMI versions.

Note: Please note that the latest version of the developer kit version with xilinx tools 2022.1 does not support cl_uram example

## Xilinx tool support

| Developer Kit Version | Tool Version Supported | Compatible FPGA Developer AMI Version       |
|-----------------------|------------------------|---------------------------------------------|
| 1.4.25+               | 2022.1                 | v1.13.X (Xilinx Vivado/Vitis 2022.1)        |
| 1.4.23+               | 2021.2                 | v1.12.X (Xilinx Vivado/Vitis 2021.2)        |
| 1.4.21+               | 2021.1                 | v1.11.X (Xilinx Vivado/Vitis 2021.1)        |
| 1.4.18+               | 2020.2                 | v1.10.X (Xilinx Vivado/Vitis 2020.2)        |
| 1.4.16+               | 2020.1                 | v1.9.0-v1.9.X (Xilinx Vivado/Vitis 2020.1)  |
| 1.4.13+               | 2019.2                 | v1.8.0-v1.8.X (Xilinx Vivado/Vitis 2019.2)  |
| 1.4.11+               | 2019.1                 | v1.7.0-v1.7.X (Xilinx Vivado/SDx 2019.1)    |
| 1.4.8 - 1.4.15b       | 2018.3                 | v1.6.0-v1.6.X (Xilinx Vivado/SDx 2018.3)    |
| 1.4.3 - 1.4.15b       | 2018.2                 | v1.5.0-v1.5.X (Xilinx Vivado/SDx 2018.2)    |
| ⚠️ 1.3.7 - 1.4.15b    | 2017.4                 | v1.4.0-v1.4.X (Xilinx Vivado/SDx 2017.4) ⚠️ |


⚠️ Developer kit release v1.4.16 will remove support for Xilinx 2017.4, 2018.2, 2018.3 toolsets. While developer kit release v1.4.16 onwards will not support older Xilinx tools, you can still use them using HDK releases v1.4.15b or earlier. 
Please check out [the latest v1.4.15b release tag from Github](https://github.com/aws/aws-fpga/releases/tag/v1.4.15b) to use Xilinx 2017.4, 2018.2, 2018.3 toolsets.

For deprecation notices, please check the [End of life announces](./README.md#end-of-life-announcements)

For software-defined development please look at the runtime compatibility table based on the Xilinx toolset in use:
[SDAccel](SDAccel/docs/Create_Runtime_AMI.md#runtime-ami-compatibility-table) or [Vitis](Vitis/docs/Create_Runtime_AMI.md#runtime-ami-compatibility-table)

### End of life Announcements

| Xilinx Tool version | State | Statement                                                                                                                                                                 | 
|-----------|-----------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 2017.1 | 🚫 Deprecated on 09/01/2018 | Developer kit versions prior to v1.3.7 and Developer AMI prior to v1.4 (2017.1) [reached end-of-life](https://forums.aws.amazon.com/ann.jspa?annID=6068).                 |
| 2017.4 | 🚫 Deprecated on 12/31/2021 | [Support for Xilinx 2017.4 toolsets was deprecated on 12/31/2021](https://forums.aws.amazon.com/ann.jspa?annID=8949). |
| 2020.1 and below | Discontinued on 02/2022 | Removed the ability for customers to newly subscribe to 2020.1 and below AMI versions to remove exposure to [CVE-2021-44228](https://www.cve.org/CVERecord?id=CVE-2021-44228) as these versions of tools do not have patches from xilinx |

## Hardware Development Kit (HDK)

The [HDK directory](./hdk/README.md) contains documentation, examples, simulation, build and AFI creation scripts to start building Amazon FPGA Images (AFI).  
The HDK can be installed on any on-premises server or an EC2 instance. 
The developer kit is not required if you plan to use a pre-built AFI shared from another developer.

### AWS Shells

With Amazon EC2 FPGA instances, each FPGA is divided into two partitions:

* Shell (SH) – AWS platform logic implementing the FPGA external peripherals, PCIe, DRAM, and Interrupts.

* Custom Logic (CL) – Custom acceleration logic created by an FPGA Developer.

At the end of the development process, combining the Shell and CL creates an Amazon FPGA Image (AFI) that can be loaded onto the Amazon EC2 FPGA Instances.

The following table provides the shells currently available to develop your CL with. Each shell provides specific interfaces and features and currently needs to be used with the Dev Kit branch listed in the table.

| Shell Name| Shell Version | Dev Kit Branch | Description|
|--------|--------|---------|-------|
| F1 XDMA Shell | F1.X.1.4 | [master](https://github.com/aws/aws-fpga/) | Provides all the [interfaces listed here](https://github.com/aws/aws-fpga/blob/master/hdk/docs/AWS_Shell_Interface_Specification.md), includes DMA | 
| F1 Small Shell | F1.S.1.0 | [small_shell](https://github.com/aws/aws-fpga/tree/small_shell) | Provides all the [interfaces listed here](https://github.com/aws/aws-fpga/blob/small_shell/hdk/docs/AWS_Shell_Interface_Specification.md). This shell does not include DMA engine and provides significant reduction in Shell resource usage. |

For more details, check the [FAQ](./FAQs.md#general-aws-fpga-shell-faqs)

## Software-defined Development Environment

The software-defined development environment allows customers to compile their C/C++/OpenCL code into the FPGA as kernels, and use OpenCL APIs to pass data to the FPGA. 
Software developers with no FPGA experience will find a familiar development experience that supercharges cloud applications.

In addition, this development environment allows for a mix of C/C++ and RTL accelerator designs into a C/C++ software based development environment. This method enables faster prototyping using C/C++ while supporting manual optimization of critical blocks within RTL. This approach is similar to optimizing time critical functions using software compiler optimization methods.

To get started with Xilinx SDAccel, review the [Software-defined development environment readme](SDAccel/README.md).
To get started with Xilinx Vitis, review the [Vitis unified development environment readme](Vitis/README.md).

## Runtime Tools (SDK)

The [SDK directory](./sdk/README.md) includes the runtime environment required to run on EC2 FPGA instances. It includes the drivers and tools to manage the AFIs that are loaded on the FPGA instance. The SDK isn't required during the AFI development process; it is only required once an AFI is loaded onto an EC2 FPGA instance. The following sdk resources are provided:
  * Linux Kernel Drivers - The developer kit includes three drivers:
    * [XDMA Driver](sdk/linux_kernel_drivers/xdma/README.md) - DMA interface to/from HDK accelerators.
  * [FPGA Libraries](sdk/userspace/fpga_libs) - APIs used by C/C++ host applications.
  * [FPGA Management Tools](sdk/userspace/fpga_mgmt_tools/README.md) - AFI management APIs for runtime loading/clearing FPGA image, gathering metrics and debug interface on the F1 instance.

# Amazon EC2 F1 Platform Features
* 1-8 Xilinx UltraScale+ VU9P based FPGA slots
* Per FPGA Slot, Interfaces available for Custom Logic(CL):
    * One x16 PCIe Gen 3 Interface
    * Four DDR4 RDIMM interfaces (72-bit with ECC, 16 GiB each; 64 GiB total)
    * AXI4 protocol support on all interfaces
* User-defined clock frequency driving all CL to Shell interfaces
* Multiple free running auxiliary clocks
* PCI-E endpoint presentation to Custom Logic(CL)
    * Management PF (physical function)
    * Application PF
* Virtual JTAG, Virtual LED, Virtual DIP Switches
* PCI-E interface between Shell(SH) and Custom Logic(CL).
    * SH to CL inbound 512-bit AXI4 interface
    * CL to SH outbound 512-bit AXI4 interface
    * Multiple 32-bit AXI-Lite buses for register access, mapped to different PCIe BARs
    * Maximum payload size set by the Shell
    * Maximum read request size set by the Shell
    * AXI4 error handling 
* DDR interface between SH and CL
    * CL to SH 512-bit AXI4 interface
    * 1 DDR controller implemented in the SH (always available)
    * 3 DDR controllers implemented in the CL (configurable number of implemented controllers allowed)

# Getting Started

### Getting familiar with AWS
If you have never used AWS before, we recommend you start with [AWS getting started training](https://aws.amazon.com/getting-started/), and focus on the basics of the [AWS EC2](https://aws.amazon.com/ec2/) and [AWS S3](https://aws.amazon.com/s3/) services. 
Understanding the fundamentals of these services will make it easier to work with AWS F1 and the FPGA Developer Kit.

FPGA Image generation and EC2 F1 instances are supported in the us-east-1 (N. Virginia), us-west-2 (Oregon), eu-west-1 (Ireland) and us-gov-west-1 ([GovCloud US](https://aws.amazon.com/govcloud-us/)) [regions](https://aws.amazon.com/about-aws/global-infrastructure/).

> ⚠️ <b>NOTE:</b> By default, your AWS Account will have an EC2 F1 Instance launch limit of 0. 
> Before using F1 instances, you will have to open a [Support Case](https://console.aws.amazon.com/support/home#/case/create) to increase the EC2 Instance limits to allow launching F1 instances.

### Setting up development environment for the first time 

You have the choice to develop on AWS EC2 using the [FPGA Developer AMI](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ) or on-premise. 

> ℹ️ <b>INFO:</b> We suggest starting with the FPGA Developer AMI with [build instances](#fpga-developer-ami) on EC2 as it has Xilinx tools and licenses setup for you to be able to quickly get into development.

> ℹ️ <b>INFO:</b> For on-premise development, you will need to have [Xilinx tools and licenses available for you to use](./docs/on_premise_licensing_help.md)

1. Start a Build Instance first to start your development. 
    > 💡 <b>TIP:</b> This instance does not have to be an F1 instance. You only require an F1 instance to run your AFI's(Amazon FPGA Image) once you have gone through your design build and AFI creation steps.
    
    > ℹ️ <b>INFO:</b> If you need to follow GUI Development flows, please checkout our [Developer Resources](./developer_resources/README.md) where we provide Step-By-Step guides to setting up a GUI Desktop.
1. Clone the [FPGA Developer Kit](https://github.com/aws/aws-fpga) on your instance.
    ```git clone https://github.com/aws/aws-fpga.git```
1. Follow the quickstarts from the next section.

### Quickstarts
Before you create your own AWS FPGA design, we recommend that you go through one of the step-by-step Quickstart guides:

| Description | Quickstart | Next Steps |
|----|----|----|
| Software Defined Accelerator Development using Xilinx Vitis | [Vitis hello_world Quickstart](Vitis/README.md) | [60+ Vitis examples](./Vitis/examples/), [Vitis Library Examples](./docs/examples/example_list.md) |
| Software Defined Accelerator Development using Xilinx SDAccel | [SDAccel hello_world Quickstart](SDAccel/README.md) | [60+ SDAccel examples](./SDAccel/examples/) |
| Custom Hardware Development(HDK) | [HDK hello_world Quickstart](hdk/README.md) | [CL to Shell and DRAM connectivity example](./hdk/cl/examples/cl_dram_dma), [Virtual Ethernet Application](./sdk/apps/virtual-ethernet) using the [Streaming Data Engine](./hdk/cl/examples/cl_sde) |
| IP Integrator/High Level Design(HLx) | [IPI hello_world Quickstart](hdk/cl/examples/cl_hello_world_hlx/README.md) | [IPI GUI Examples](hdk/docs/IPI_GUI_Examples.md) |

ℹ️ <b>INFO:</b> For more in-depth applications and examples of using High level synthesis, Vitis Libraries, App Notes and Workshops, please refer to our [Example List](./docs/examples/example_list.md)

### How Tos
| How To                                                                                | Description                                                                            | 

RELEASE_NOTES.md
# AWS EC2 FPGA HDK+SDK Release Notes

**NOTE:** See [ERRATA](./ERRATA.md) for unsupported features

## Release 1.4.25
* FPGA developer kit now supports Xilinx Vivado/Vitis 2022.1
* Please note the below for this version of developer kit
    * cl_uram example is currently not supported with 2022.1 version
    * When loading a vitis design on an instance running HDK AFI/design, instance needs to be rebooted before loading the vitis design
    * non-root access is unstable

## Release 1.4.24
* AWS supports [migration](./hdk/docs/U200_to_F1_migration_HDK.md) from U200 Vivado development flow to F1 HDK flow using F1.A.1.4 shell

## Release 1.4.23
* FPGA developer kit now supports Xilinx Vivado/Vitis 2021.2

## Release 1.4.22
* FPGA developer kit update to upgrade Virtual Ethernet to support jumbo frames using newer versions of dpdk/pktgen

## Release 1.4.21
* FPGA developer kit now supports Xilinx Vivado/Vitis 2021.1

## Release 1.4.20
* Bug Fix release to fix XDMA to fix Issue #516
* Miscellaneous documentation updates

## Release 1.4.19
* Bug Fix release

We have identified a bug in the `flop_ccf.sv` module that can potentially impact timing closure of designs.
The module is instantiated in `sh_ddr.sv` and inadvertently introduces a timing path on the reset logic.
Although there is no functional impact, it may increase Vivado tool’s effort in timing closure of design.
There should be no functional impact from this bug if your design has already met timing.

## Release 1.4.18
* FPGA developer kit now supports Xilinx Vivado/Vitis 2020.2

## Release 1.4.17
* Updated XDMA Driver to allow builds on newer kernels
* Updated documentation on Alveo U200 to F1 platform porting
* Added Vitis 2019.2 Patching for AR#73068

## Release 1.4.16
* FPGA developer kit now supports Xilinx Vivado/Vitis 2020.1
    * To upgrade, use [Developer AMI v1.9.0](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ) on the AWS Marketplace.
* Updated Vitis examples to include usage of Vitis Libraries.
* Added documentation and examples to show Xilinx Alveo design migration to F1.

## Release 1.4.15a
* Fixed Xilinx AR#73068 patching
    * DDR4 IP needs to be regenerated for the patch to take effect.
* Updated cl_dram_dma public AFI.

## Release 1.4.15
* Added Xilinx AR#73068 patching
* Added DMA range error to the interrupt status register metrics
* Enhanced DDR model rebuild qualifiers in hdk_setup.sh 
* Updated Virtual JTAG Documentation

## Release 1.4.14
* Updated Vitis Platform file to fix a DDR bandwidth issue
* Added Vitis Debug Documentation

## Release 1.4.13
* FPGA developer kit now supports Xilinx Vivado/Vitis 2019.2
* To upgrade, use [Developer AMI v1.8.0](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ) on the AWS Marketplace.
       
## Release 1.4.12
* Added supported versions for BJS AMI's
* Added link to the re:Invent 19 F1 workshop
* Fixed missing extern C declaration by PR #473 
* Documentation Path fixes from PR #466, #468 and #470 

## Release 1.4.11
* FPGA developer kit supports Xilinx SDx/Vivado 2019.1
    * We recommend developers upgrade to v1.4.11 to benefit from the new features, bug fixes, and optimizations. 
    * To upgrade, use [Developer AMI v1.7.0](https://aws.amazon.com/marketplace/pp/B06VVYBLZZ) on the AWS Marketplace. The Developer Kit scripts (hdk_setup.sh or sdaccel_setup.sh) will detect the tool version and update the environment based on requirements needed for Xilinx 2019.1 tools. 
* New functionality:
    * Added a [developer resources section](./developer_resources/README.md) that provides guides on how to setup your own GUI Desktop and compute cluster environment.
    * Developers can now ask for AFI limit increases via the [AWS Support Center Console](https://console.aws.amazon.com/support/cases#/create).
        * Create a case to increase your `EC2 FPGA` service limit from the console.
    * HLx IPI flow updates
        * HLx support for AXI Fast Memory mode.
        * HLx support for 3rd party simulations.
        * HLx support for changes in shell and AWS IP updates(e.g. sh_ddr).     
* Bug Fixes:
    * Documentation fixes in the [Shell Interface Specification](./hdk/docs/AWS_Shell_Interface_Specification.md)
    * Fixes for forum questions
        * [Unable to compile aws_v1_0_vl_rfs.sv in Synopsys VCS](https://forums.aws.amazon.com/thread.jspa?threadID=308829&tstart=0)
        * [Use fpga_mgmt init in HLx runtime](https://forums.aws.amazon.com/thread.jspa?messageID=912063)
    * New XRT versions added to the [XRT Installation Instructions](./SDAccel/docs/XRT_installation_instructions.md) to fix segmentation faults when using xclbin instead of awsxclbin files.
* Deprecations:
    * Removed GUI Setup scripts from AMI v1.7.0 onwards. See the [developer resources section](./developer_resources/README.md) that provides guides on how to setup your own GUI Desktop and compute cluster environment.
* Package versions used for validation
  
   | Package | AMI 1.7.0 [2019.1] | AMI 1.6.0 [2018.3] |AMI 1.5.0 [2018.2] | AMI 1.4.0 [2017.4] |
   |---------|---|------------------------|------------------------|-----------------------|
   | OS      | Centos 7.6 | Centos 7.6 | Centos 7.5, 7.6 | Centos 7.4 |
   | kernel  | 3.10.0-957.27.2.el7.x86_64 | 3.10.0-957.5.1.el7.x86_64 | 3.10.0-862.11.6.el7.x86_64, 3.10.0-957.1.3.el7.x86_64  | 3.10.0-693.21.1.el7.x86_64 |
   | kernel-devel | 3.10.0-957.27.2.el7.x86_64 | 3.10.0-957.5.1.el7.x86_64 | 3.10.0-862.11.6.el7.x86_64, 3.10.0-957.1.3.el7.x86_64 | 3.10.0-693.21.1.el7.x86_64 |
   | LIBSTDC++ | libstdc++-4.8.5-36.el7_6.2.x86_64 | libstdc++-4.8.5-36.el7.x86_64 | libstdc++-4.8.5-36.el7.x86_64 | libstdc++-4.8.5-16.el7_4.2.x86_64 |   

## Release 1.4.10
* New functionality:
    * SDK now sorts the slots in DBDF order. Any scripts or integration maintainers should note that the slot order will be different from previous versions and should make any updates accordingly.
  
* Bug Fixes:

vitis_setup.sh
# Amazon FPGA Hardware Development Kit
#
# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Amazon Software License (the "License"). You may not use
# this file except in compliance with the License. A copy of the License is
# located at
#
#    http://aws.amazon.com/asl/
#
# or in the "license" file accompanying this file. This file is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or
# implied. See the License for the specific language governing permissions and
# limitations under the License.

# Script must be sourced from a bash shell or it will not work
# When being sourced $0 will be the interactive shell and $BASH_SOURCE_ will contain the script being sourced
# When being run $0 and $_ will be the same.
script=${BASH_SOURCE[0]}
if [ $script == $0 ]; then
    echo "ERROR: You must source this script"
    exit 2
fi

full_script=$(readlink -f $script)
script_name=$(basename $full_script)
script_dir=$(dirname $full_script)
current_dir=$(pwd)

source $script_dir/shared/bin/set_common_functions.sh
source $script_dir/shared/bin/set_common_env_vars.sh
source $VITIS_DIR/Runtime/xrt_common_functions.sh

# Source sdk_setup.sh
info_msg "Sourcing sdk_setup.sh"
if ! source $AWS_FPGA_REPO_DIR/sdk_setup.sh; then
    return 1
fi

if [ -z "$SDK_DIR" ]; then
    err_msg "SDK_DIR environment variable is not set.  Please use 'source sdk_setup.sh' from the aws-fpga directory."
    return 1
fi

debug=0

function usage {
    echo -e "USAGE: source [\$AWS_FPGA_REPO_DIR/]$script_name [-d|-debug] [-h|-help]"
}

function help {
    info_msg "$script_name"
    info_msg " "
    info_msg "Sets up the environment for AWS FPGA Vitis tools."
    info_msg " "
    info_msg "vitis_setup.sh script will:"
    info_msg "  (1) install FPGA Management Tools,"
    info_msg "  (2) check if Xilinx tools are available,"
    info_msg "  (3) check if required packages are installed,"
    info_msg "  (4) Download lastest AWS Platform,"
    info_msg "  (5) Install Runtime drivers "
    echo " "
    usage
}

function check_set_xilinx_vitis {
    if [[ ":$XILINX_VITIS" == ':' ]]; then
        debug_msg "XILINX_VITIS is not set"
        which vitis
        RET=$?
        if [ $RET != 0 ]; then
            debug_msg "vitis not found in path."
            err_msg "XILINX_VITIS variable not set and vitis not in the path"
            err_msg "Please set XILINX_VITIS variable to point to your location of your Xilinx installation or add location of vitis exectuable to your PATH variable"
            return $RET
        else
            export XILINX_VITIS=`which vitis | sed 's:/bin/vitis::'`
            info_msg "Setting XILINX_VITIS to $XILINX_VITIS"
        fi
    else
        info_msg "XILINX_VITIS is already set to $XILINX_VITIS"
    fi
    # get Vitis release version, i.e. "2019.2"
    RELEASE_VER=$(basename $XILINX_VITIS)
    RELEASE_VER=${RELEASE_VER:0:6}
    export RELEASE_VER=$RELEASE_VER
    echo "RELEASE_VER equals $RELEASE_VER"
}

function check_install_packages_centos {
#TODO: Check required packages are installed or install them
#TODO: Check version of gcc is above 4.8.5 (4.6.3 does not work)
  for pkg in `cat $VITIS_DIR/packages.txt`; do
    if yum list installed "$pkg" >/dev/null 2>&1; then
      true
    else
      warn_msg " $pkg not installed - please run: sudo yum install $pkg "
    fi
  done
}

function check_install_packages_ubuntu {
  for pkg in `cat $VITIS_DIR/packages.txt`; do
    if apt -qq list "$pkg" >/dev/null 2>&1; then
      true
    else
      warn_msg " $pkg not installed - please run: sudo apt-get install $pkg "
    fi
  done
}

function check_internet {
    curl --silent --head -m 30 http://www.amazon.com
    RET=$?
    if [ $RET != 0 ]; then
        err_msg "curl cannot connect to the internet using please check your internet connection or proxy settings"
        err_msg "To check your connection run:   curl --silent --head -m 30 http://www.amazon.com "
        return $RET
    else
        info_msg "Internet Access OK"
    fi
}

function check_icd {
    info_msg "Checking ICD is installed"
    if grep -q 'libxilinxopencl.so' /etc/OpenCL/vendors/xilinx.icd; then
        info_msg "Found 'libxilinxopencl.so"
    else
        info_msg "/etc/OpenCL/vendors/xilinx.icd does not exist or does not contain lbixilinxopencl.so creating and adding libxilinxopencl.so to it"
        sudo sh -c "echo libxilinxopencl.so > /etc/OpenCL/vendors/xilinx.icd"
        RET=$?
    if [ $RET != 0 ]; then
        err_msg "/etc/OpenCL/vendors/xilinx.icd does not exist and cannot be created, sudo permissions needed to update it."
        err_msg "Run the following with sudo permissions: sudo sh -c \"echo libxilinxopencl.so > /etc/OpenCL/vendors/xilinx.icd\" "
        return $RET
    else
        echo "Done with ICD installation"
    fi
    fi
}

# Process command line args
args=( "$@" )
for (( i = 0; i < ${#args[@]}; i++ )); do
    arg=${args[$i]}
    case $arg in
        -d|-debug)
            debug=1
        ;;
        -h|-help)
            help
            return 0
        ;;
        *)
            err_msg "Invalid option: $arg\n"
            usage
            return 1
    esac
done

# Check XILINX_VITIS is set
if ! check_set_xilinx_vitis; then
    return 1
fi

info_msg " XILINX_VITIS is set to $XILINX_VITIS"
# Install patches as required.
info_msg "Setting up Vitis patches if required."
setup_patches


# Update Xilinx Vitis Examples from GitHub
info_msg "Using Vitis $RELEASE_VER"
if [[ $RELEASE_VER =~ .*2019\.2.*  ||  $RELEASE_VER =~ .*2020\.* ||  $RELEASE_VER =~ .*2021\.* ||  $RELEASE_VER =~ .*2022\.1.* ]]; then
    info_msg "Updating Xilinx Vitis Examples $RELEASE_VER"
    git submodule update --init -- Vitis/examples/xilinx_$RELEASE_VER
    export VIVADO_TOOL_VER=$RELEASE_VER
    if [ -e $VITIS_DIR/examples/xilinx ]; then
        if [ ! -L $VITIS_DIR/examples/xilinx ]; then
          err_msg "ERROR:  Vitis/examples/xilinx is not a symbolic link.  Backup any data and remove Vitis/examples/xilinx directory.  The setup needs to create a symbolic link from Vitis/examples/xilinx to Vitis/examples/xilinx_$RELEASE_VER"
          return 1
        fi
    fi
    ln -sf $VITIS_DIR/examples/xilinx_$RELEASE_VER $VITIS_DIR/examples/xilinx
else
   echo " $RELEASE_VER is not supported (2019.2, 2020.1, 2020.2, 2021.1, 2021.2 or 2022.1 are supported).\n"
   return 2
fi

# Check if internet connection is available
if ! check_internet; then
    return 1
fi

# Check ICD is installed
if ! check_icd; then
    return 1
fi

# Check correct packages are installed

if [[ $(lsb_release -si) == "Centos" ]]; then
    if ! check_install_packages_centos; then
        return 1
    fi
elif [[ $(lsb_release -si) == "Ubuntu" ]]; then
    if ! check_install_packages_ubuntu; then
        return 1
    fi
#elif [[ $(lsb_release -si) == "Ubuntu" ]]; then
#    if ! check_install_packages_ubuntu; then
#        return 1
#    fi
fi

function setup_xsa {

    if [ "$#" -ne 3 ]; then
        err_msg "Illegal number of parameters sent to the setup_xsa function!"
        return 1
    fi

    XSA=$1
    XSA_S3_BASE_DIR=$2
    PLATFORM_ENV_VAR_NAME=$3

    xsa_dir=$VITIS_DIR/aws_platform/$XSA/hw/
    vitis_xsa=$xsa_dir/$XSA.xsa
    vitis_xsa_s3_bucket=aws-fpga-hdk-resources
    s3_vitis_xsa=$vitis_xsa_s3_bucket/Vitis/$XSA_S3_BASE_DIR/$XSA/$XSA.xsa

    # set a variable to point to the platform for build and emulation runs
    export "$PLATFORM_ENV_VAR_NAME"=$VITIS_DIR/aws_platform/$XSA/$XSA.xpfm

    # Download the sha256
    if [ ! -e $xsa_dir ]; then
        mkdir -p $xsa_dir || { err_msg "Failed to create $xsa_dir"; return 2; }
    fi

    # Use curl instead of AWS CLI so that credentials aren't required.
    curl -s https://s3.amazonaws.com/$s3_vitis_xsa.sha256 -o $vitis_xsa.sha256 || { err_msg "Failed to download XSA version from $s3_vitis_xsa.sha256 -o $vitis_xsa.sha256"; return 2; }
    if grep -q '<?xml version' $vitis_xsa.sha256; then
        err_msg "Failed to download VITIS XSA version from $s3_vitis_xsa.sha256"
        cat vitis_xsa.sha256
        return 2
    fi
    exp_sha256=$(cat $vitis_xsa.sha256)
    debug_msg "  latest   version=$exp_sha256"
    # If XSA already downloaded check its sha256
    if [ -e $vitis_xsa ]; then
        act_sha256=$( sha256sum $vitis_xsa | awk '{ print $1 }' )
        debug_msg "  existing version=$act_sha256"
        if [[ $act_sha256 != $exp_sha256 ]]; then
            info_msg "VITIS XSA version is incorrect"
            info_msg "  Saving old XSA to $vitis_xsa.back"
            mv $vitis_xsa $vitis_xsa.back
        fi
    else
        info_msg "VITIS XSA hasn't been downloaded yet."
    fi

    if [ ! -e $vitis_xsa ]; then
        info_msg "Downloading latest VITIS XSA from $s3_vitis_xsa"
        # Use curl instead of AWS CLI so that credentials aren't required.
        curl -s https://s3.amazonaws.com/$s3_vitis_xsa -o $vitis_xsa || { err_msg "VITIS XSA download failed"; return 2; }
    fi
    # Check sha256
    act_sha256=$( sha256sum $vitis_xsa | awk '{ print $1 }' )
    if [[ $act_sha256 != $exp_sha256 ]]; then
        err_msg "Incorrect VITIS XSA version:"
        err_msg "  expected version=$exp_sha256"
        err_msg "  actual   version=$act_sha256"
        err_msg "  There may be an issue with the uploaded XSA or the download failed."
        return 2
    fi
}

    #-------------------201920_3 Vitis Platform----------------------
    setup_xsa xilinx_aws-vu9p-f1_shell-v04261818_201920_3 xsa_v121319_shell_v04261818 AWS_PLATFORM_201920_3
    info_msg "AWS Platform: 201920_3 Vitis Platform is up-to-date"
    #-------------------201920_3 Vitis Platform----------------------


# Setup XRT as we need it for building
setup_runtime

export AWS_PLATFORM=$AWS_PLATFORM_201920_3
info_msg "The default AWS Platform has been set to: \"AWS_PLATFORM=\$AWS_PLATFORM_201920_3\" "

info_msg "Vitis Setup PASSED"

info_msg "To run a runtime example, start the MPD service by calling: \`systemctl is-active --quiet mpd || sudo systemctl start mpd\`"
